add config options to allow global coord clamping

Arkessler opened the pull request:
- allows one to set a minimum and maximum x and y to act as boundaries for the camera projection
- works for both dragging and zooming

Fixes: #11

Arkessler commented:
Hello! I've been getting into bevy recently, and I found your camera super helpful. Thank you for open sourcing it! I did find, however, that I wanted to be able to restrict the camera's view window so that I didn't drag or zoom the window beyond the borders of a specific grid. I did some local tinkering and I got it working pretty much as I envisioned. If the change looks good to you, I thought it'd be a nice first contribution to the bevy community to merge this support upstream :) 

Video of the implemented behavior:
https://user-images.githubusercontent.com/8944532/189548695-7bd62198-eea7-4fa0-a576-6d1f01b02a2b.mp4

johanhelsing commented on readme.md:
````diff
@@ -48,6 +49,22 @@ commands.spawn_bundle(Camera2dBundle::default())
     });
 ```
 
+Clamped camera example
````
Could you move this to a separate example in the examples folder? Having them as actual examples makes it easier for people to run them, and also makes it easier to maintain, since they are build tested in ci.

johanhelsing commented on src/lib.rs:
```diff
@@ -142,6 +232,26 @@ pub struct PanCam {
     /// If present, the orthographic projection's scale will be clamped at
     /// this value when zooming out.
     pub max_scale: Option<f32>,
+    /// The minimum global x of the camera window. Uses global pixel positions
```
Don't these use local positions not global positions (Transform vs GlobalTransform)? What happens if the camera is parented under another transform? For what it's worth, I'd rather they used local coordinates than global, so if I'm not mistaken, I'd rather keep the behavior and change the documentation/naming (`min_x`).

Also, I feel calling the units pixels is probably also misleading, that's only true for specific scales and many applications don't have a concept of pixels either. Maybe just keep it agnostic and call it `minimum x position of the camera window`?

johanhelsing commented on src/lib.rs:
```diff
@@ -56,10 +56,41 @@ fn camera_zoom(
 
+            // If there is both a min and max x, that limits how far we can zoom. Make sure we don't exceed that
+            if let (Some(min_global_x), Some(max_global_x)) = (cam.min_global_x, cam.max_global_x) {
+                // Arkessler: I was not smart enough to solve this on my own, I plugged the combination of the scaling
```
If possible, I'd like us to try and make the math as readable as possible, so we can remove the comments and let the code be mostly self-documenting.

johanhelsing commented on src/lib.rs:
```diff
@@ -69,6 +100,36 @@ fn camera_zoom(
+                if let Some(min_global_y) = cam.min_global_y {
+                    // Use the global pixel position provided to calculate the minimum camera y position allowable
+                    let min_cam_y = min_global_y + ((window.height() / 2.) * scaling.y);
+                    // Override the transform with the limit if needed. This prevents weird behavior with large mouse deltas
```
This is to prevent us from zooming out over the border so it becomes visible, right? Could you maybe change the comment to explain that?

johanhelsing commented on src/lib.rs:
```diff
@@ -69,6 +100,36 @@ fn camera_zoom(
+                    window.height() / (proj.top - proj.bottom),
+                ) * proj.scale;
+                if let Some(min_global_x) = cam.min_global_x {
+                    // Use the global pixel position provided to calculate the minimum camera x position allowable
```
I think these comments are redundant. The code is clear on its own.

johanhelsing commented:
Hey, and thanks for the contribution! It's a difficult task with lots of *literal* edge cases :)

Overall, it looks like the implementation makes sense, but like you say, it's kind of hard to wrap your head around the equation as it stands, and it adds quite a lot of complexity to otherwise straightforward code. If possible I'd like us to try and make it more readable.

johanhelsing commented on src/lib.rs:
```diff
@@ -56,10 +56,41 @@ fn camera_zoom(
+                // solve for s (i + ((w / 2) * ((w / (r - l) * s)))) = (j - ((w / 2) * ((w / (r - l) * s))))
+                let max_global_x_scale = (((max_global_x - min_global_x)
+                    * (proj.left - proj.right))
+                    / window.width().powi(2))
```
It feels like this would be more readable if arranged like this:

```rust
let cam_projection_width = proj.right - proj.left;
let max_window_width = max_global_x - min_global_x;
let max_global_x_scale = max_window_width / window.width() * cam_projection_width / window.width();
```

Still not sure I completely understand it, though.

Arkessler commented:
@johanhelsing Apologies, wasn't able to dedicate enough heads-down time dig into the zooming math until the weekend. I'm still not sure I can explain how it works exhaustively, but I've tried to clean up the math to make it more readable (your suggestions were incredibly helpful in doing so!). 

I moved the max allowable zoom scale logic to two new functions, `max_scale_within_x_bounds` and `max_scale_within_y_bounds`, mostly so that I could add commentary explaining what was going on without it bloating the original function. I'm happy to remove the long function comment if it feels redundant, but I found it personally helpful to write out a number of examples of how the projection, window, and boundary dimensions would interact with one another. Kind of an "if you turn these knobs, this is what will happen" set.

I renamed all the global/pixel stuff to remove the misleading elements. I think I made a lot of bad assumptions when first trying to understand how cameras work in bevy. I'm definitely not to full understanding yet, but I think I'm starting to grok the basic components a bit better.

I also moved the example to the examples folder, I hadn't even known you could execute examples via `cargo run --example`, cargo is pretty cool! I appreciate your patience and helpful suggestions, I feel like I'm learning a lot about how to write Rust and Bevy 😄 If you wouldn't mind taking a look at the updated code, I'm of course happy to make any others changes that jump out.

johanhelsing commented on src/lib.rs:
```diff
@@ -69,11 +84,112 @@ fn camera_zoom(
+) -> f32 {
+    let projection_width = proj.right - proj.left;
+    let default_projection_scale = projection_width / window_width;
+
```
I've been kind of slow to respond because I've been on the go and haven't really had the time to sit down and look at this closely.

I feel like this is really close. With the right kind of comments about the units of measure of the values it would probably make sense

Also, OrthographicProjection has a scaling mode enum. I'm not sure if we have to account for that here or not?

johanhelsing replied:
So the scaling_mode just determines what the proj.left right etc. will be set to, so we don't need to think about that.

johanhelsing replied:
https://github.com/Arkessler/bevy_pancam/pull/1

Arkessler replied:
Sorry for being slow to get back, it can be hard to find time for side projects when balancing work and home life. But wow, I feel like you cracked the final piece of this with the simplification of the math. Nicely done! I merged your PR to my fork, and am going to see if I can get some unit tests in place. I'm still wrapping my head a bit around Rust unit testing, but I think it's a good opportunity to learn it while also making sure some of my assumptions are correct

johanhelsing reviewed (commented):
Good to know you're appreciating the reviews! It's often hard to tell whether people enjoy it, or just want to get their pr merged as quickly as possible. We could merge this as is, and i could look over the math later, it's up to you how much time you want to put into it.

Thanks again for working on this!

johanhelsing commented on src/lib.rs:
```diff
@@ -69,11 +84,112 @@ fn camera_zoom(
+///         max_safe_x_scale = 0.5 * 3. = 1.5
+///         (Scrolling out this far lets us scroll beyond the original image and see empty space on either side)
+fn max_scale_within_x_bounds(
+    min_x_bound: &f32,
```
Pass `f32`s by value, not reference.

johanhelsing commented on src/lib.rs:
```diff
@@ -69,11 +84,112 @@ fn camera_zoom(
+/// The examples below aim to explain why we need to account for both, and what happens when we turn the
+/// various knobs
+///
+/// - In the simplest case where the width of the projection and window are both 100, and assuming we
```
For example:

```rust
#[cfg(test)]
mod test {
    use super::*;
    use bevy::{prelude::OrthographicProjection, render::camera::CameraProjection};

    #[test]
    fn simple_clamping() {
        // In the simplest case where the width of the projection and window are the same
        // then we should be unable to scroll further out than 1, which would show the entire window
        let mut proj = OrthographicProjection::default();
        let window_width = 800.;
        let window_height = 600.;
        proj.update(window_width, window_height);
        let max_safe_x_scale = max_scale_within_x_bounds(0., window_width, window_width, &proj);
        assert_eq!(max_safe_x_scale, 1.);
    }
}
```

johanhelsing replied:
I was able to simplify the math. See PR to this branch https://github.com/Arkessler/bevy_pancam/pull/1

With that we probably don't need tests or extensive explanation, it's pretty straightforward.

johanhelsing commented on src/lib.rs:
```diff
@@ -69,11 +84,112 @@ fn camera_zoom(
+    min_x_bound: &f32,
+    max_x_bound: &f32,
+    window_width: &f32,
+    proj: &OrthographicProjection,
```
See: https://github.com/Arkessler/bevy_pancam/pull/1

Arkessler commented:
Thank you for the PR @johanhelsing , I think that really provided the remaining refinement needed on the scaling functions! I just put up another commit to add some unit tests for the scale functions and hopefully make the `clamped_border` comments more useful. I think I'm starting to feel pretty good about merging this (assuming it passes CI), but as always am happy to address anything you think might be remaining

johanhelsing commented:
> Sorry for being slow to get back, it can be hard to find time for side projects when balancing work and home life. But wow, I feel like you cracked the final piece of this with the simplification of the math. Nicely done! I merged your PR to my fork, and am going to see if I can get some unit tests in place. I'm still wrapping my head a bit around Rust unit testing, but I think it's a good opportunity to learn it while also making sure some of my assumptions are correct

No worries. I've also been juggling a lot lately.

I'm happy it turned out to be really simple and readable in the end :) Thanks again!
